support-frontend/app/controllers/Application.scala

```diff
@@ -274,15 +281,29 @@ class Application(
   }
 
   def router(countryGroupId: String): Action[AnyContent] = MaybeAuthenticatedAction { implicit request =>
-    implicit val settings: AllSettings = settingsProvider.getAllSettings()
+    request.queryString
+      .getOrElse("product", Nil)
+      .headOption
+      .flatMap(productString => Product.fromString(productString))
+      .map(routeForProduct(_))
+      .getOrElse(BadRequest("No product name provided"))
+  }
 
+  def routeForProduct(product: Product)(implicit request: OptionalAuthRequest[AnyContent]) = {
+    implicit val settings: AllSettings = settingsProvider.getAllSettings()
     val geoData = request.geoData
     val serversideTests = generateParticipations(Nil)
     val isTestUser = testUserService.isTestUser(request)
     // This will be present if the token has been flashed into the session by the PayPal redirect endpoint
     val guestAccountCreationToken = request.flash.get("guestAccountCreationToken")
     val productCatalog = cachedProductCatalogServiceProvider.fromStage(stage, isTestUser).get()
 
+    val queryPromos =
+      request.queryString
+        .getOrElse("promoCode", Nil)
+        .toList
+    val productPrices = priceSummaryServiceProvider.forUser(isTestUser).getPrices(product, queryPromos)
```

Comment:
To get your thoughts on this, I was trying to avoid using this model as:
- it is an extra layer of transformation / indirection between the source data and the app (we have traversed through this logic a lot while pairing when working on promotions)
- it needs context switching between the scala app and the JS app
- it's using the old Zuora catalog as source data

Maybe as a follow up we might look at how we pass down promo logic in the same shape as the source data, but validate promoCodes on the server, as I guess this is part of why this service exists?

Reply:
Yes I totally agree that we should move beyond using this approach but I didn't want to guess at how you were planning on doing that and this seemed the easiest way to get promotions going for now.